lib/client/ipam_block_reader_writer.go

```diff
@@ -310,23 +311,32 @@ func randomBlockGenerator(pool cnet.IPNet) func() *cnet.IPNet {
 	prefixLen := size - ones
 	numBlocks := int(math.Exp2(float64(prefixLen))) / blockSize
 
-	// Generate a randomly ordered slice of block indexes.
+	// Start at a random offset index
 	source := rand.NewSource(time.Now().UnixNano())
 	randm := rand.New(source)
-	idxs := randm.Perm(numBlocks)
-	i := 0
+
+	// initialIndex keeps track of the random starting point
+	initialIndex := randm.Intn(numBlocks)
+	// i keeps track of current index while walking the blocks in a pool
+	i := initialIndex
+	numReturned := 0
+
 	return func() *cnet.IPNet {
-		// Pop a block index off of the indexes slice.
-		if len(idxs) != 0 {
-			i, idxs = idxs[len(idxs)-1], idxs[:len(idxs)-1]
-		} else {
+		ip := incrementIP(baseIP, i*blockSize)
+		ipnet := net.IPNet{ip.IP, version.BlockPrefixMask}
+		i++
+		if i >= numBlocks {
```

Comment:
I think this would be clearer like this:

```go
if i >= numBlocks - 1 {
  // Loop back to beginning of pool rather than 
  // increment, because incrementing would put us outside of the pool.
  i = 0
} else {
  // Increment to next block.
  i++
}
```